### DISCUSSION TITLE
Do yall need help adapting this project to be a media player?

### DISCUSSION THREAD

**Discussion by nednoodlehead** [python-ffmpegio/python-ffmpegio#6] [General]
I'm not too well versed with like audio-specific manipulation, but if you get me a function that I can play audio I can make an audio player out of it. 
This is mostly a response to [Kesh](https://stackoverflow.com/questions/73524101/is-there-a-fix-for-the-flashing-console-bug-with-ffmpeg?noredirect=1#comment129837073_73524101) on StackOverflow

I made one before using PyDub bindings (Which uses SimpleAudio, which uses FFMPEG). It is a very convoluted process, and making a binding through here would be significantly easier.

The current project that I'm using my audioplayer in is [Punge](https://github.com/nednoodlehead/punge).
A large part of the gimmick that makes the audio player work is that the music does not have an inherent 'time.sleep()' type of attribute, so if you call "pydub.playback._play_with_simpleaudio(audiosegment)", it won't do anything. But if you call a time.sleep() along with it, it will play for the duration of the time.sleep()

There are also two limitations that I've encountered; I'm unsure of this is a hole in the package I'm using, or I haven't figured out a way around (I'll admit it, I haven't tried too much): Scrubbing and count of length of song & time left.

The current player I have is fairly bugless but has two main issues that may be solved by using a new library (like this one):
1). It is slightly clunky and for long videos (15m>), it can take a few second to initialize.
2). Memory usage for the app directly correlates with the length of the current video. 50mb average, and up to 150mb for longer videos

Let me know what you guys think, and again, I would probably need binds for playing audio, as I don't know much about how audio specifically works.

**Comment by tikuma-lsuhsc** [comment]
@nednoodlehead - I must have missed your response to my comment on SO. Sorry 'bout that.

Anyway, back to my comment, this project itself won't turn into a GUI app, but I'm planning on using it for a future GUI project. So, if calling ffmpeg subprocess causes a glitch, I'd like to have a way to avoid it.

> 1). Pydub is used to read data, turn it into an 'audiosegment', then call pydub.playback._play_with_simpleaudio on it (Which relies on ffmpeg). I could try your package, see how it works but I have written 400 lines to handle how the audiosegment is handled, so minor inconvenience. 

I'll take a look at your project later this weekend, to see if maybe the `ffmpegio`'s stream reader can help your project.

**Comment by tikuma-lsuhsc** [comment]
Well, you got me curious enough to test what I designed `ffmpegio` to do and it indeed does work as advertised.

Here is a quick demo on how you can load data while playing (need `ffmpegio-core` package only). `simpleaudio` is consuming the audio samples in the `audio_data` buffer while the script is filling the buffer. I'm curious if this alleviates your issue #1.

Because I don't use a custom class, I think it's fairly easy to program with `ffmpegio`. Check it out.

```python
import ffmpegio
import simpleaudio as sa
from pprint import pprint

file = "path_to_audio_or_video_file.mp4"

# grab the file info
info = ffmpegio.probe.full_details(file)
sinfo = next((s for s in info["streams"] if s["codec_type"] == "audio"))
pprint(sinfo)

sample_rate = int(sinfo["sample_rate"])
num_channels = 2  # int(sinfo["channels"]) # force stereo
sample_fmt = "s16"  # force signed 16-bit (simpleaudio only supports integer-multiple byte formats)
bytes_per_sample = 2  # s16 = 2 bytes/sample

# estimate the number of samples (try stream duration first, if not available use container duration)
nb_samples = int(
    float(sinfo.get("duration", info["format"].get("duration")) * sample_rate)
)

# preallocate the buffer
ntotal = nb_samples * num_channels  # total data size in number of bytes
audio_data = bytearray(ntotal)  # playback buffer

# open ffmpegio's stream-reader to read nblk samples at a time
nblk = sample_rate  # 1 second (you can use any number you wish)
with ffmpegio.open(
    file, "ra", sample_fmt=sample_fmt, ac=num_channels, blocksize=nblk
) as f:

    # read the first block of data and place them on the playback buffer
    data = f.read(nblk)["buffer"]
    nwritten = len(data)
    audio_data[:nwritten] = data

    # start the audio player (before finish loading the data)
    play_obj = sa.play_buffer(audio_data, num_channels, bytes_per_sample, sample_rate)

    # queue up rest of the data while playing
    for block in f:  # process one block at a time
        # make sure not to overflow the playback buffer
        data = block["buffer"]
        n = len(data)
        nend = min(nwritten + n, ntotal)
        audio_data[nwritten:nend] = data[: nend - nwritten]
        nwritten = nend  # update bytes written
```

**Reply by nednoodlehead** [reply]
Running that code using either mp3, or wav gives an error on line 33:
IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

Also, the [Punge](https://github.com/nednoodlehead/punge/blob/master/punge_main.py) project, has a lot of lines, the class you're looking for is from line 447 -> 957.

I'm also certain this will alleviate the memory issue, for the scrubbing, I think it will be possible as well. I'll look at implementing this once:
1). The code doesn't give me an error lol
2). I understand what's going on

My only question here is, is crossfade for audio on the table still? One of the main reasons I chose Pydub was because it supported crossfade operations. Would something like that be possible, or even better, easy?

**Reply by tikuma-lsuhsc** [reply]
That's because you're using the full `ffmpegio` version, which uses `numpy` arrays for data transaction. The above code uses only `ffmpegio-core` which uses `dict`s with `bytes` data. Uninstall `ffmpegio` and keep `ffmpegio-core` and see what happens.

**Reply by nednoodlehead** [reply]
>pip uninstall ffmpegio
>pip install ffmpegio-core

Is there another way im meant to do this? cause doing these just produces the 
"ModuleNotFoundError: No module named 'ffmpegio'"

**Reply by tikuma-lsuhsc** [reply]
That's strange... If you `pip list` do they appear?

BTW, here's an [GitHub repo](https://github.com/python-ffmpegio/python-ffmpegio-example-audioplayer) just created for the code above.

**Reply by nednoodlehead** [reply]
Pip list gives (after uninstalling ffmpegio):
![image](https://user-images.githubusercontent.com/102321690/188246675-ce716f48-6d45-4d24-9b56-6822eaf5bb78.png)
Pip list gives (after re-installing ffmpegio):
![image](https://user-images.githubusercontent.com/102321690/188246714-185474e8-2013-4690-9213-0b2f8cbdeaba.png)

I ran the code on each of them, both gave the same index error listed above. (The index error)

I would assume that this is probably something on my end, but I tested in different directories, different files, and it seems to give the same error

**Reply by tikuma-lsuhsc** [reply]
Oh, you don't need `ffmpegio` package. `ffmpegio-core` package is all you need. `ffmpegio` package is actually installing `ffmpegio-plugin-numpy` plugin module. Yeah, it's confusing but it's done this way for a historical reason. At v0.3.0, I split `ffmpegio-core` from `ffmpegio` to remove the `numpy` dependency.

**Reply by nednoodlehead** [reply]
Solved by creating a new venv. Dont know why it was mad at me though

**Comment by tikuma-lsuhsc** [comment]
> crossfade

I'm curious to see `pydub` does it. It can be done, but not trivial to me atm (you can do fade in & out easily)

**Reply by nednoodlehead** [reply]
Yeah fading in and out is a common operation of most audio projects. Chaining them together is a totally different story though.
In Pydub, you could easily stitch two AudioSegments together and use a .crossfade() or whatever one them, but if you wanted it to happen to many audiosegments, seamlessly, it is a lot harder.

Because you couldn't just continually append them together without some type of 'cut' between them when you switch to a different AudioSegment (crossfading would combine two together).

A helpful part of the no 'time.sleep()' stuff by default, is that this operation becomes slightly easier. Instead of messing around with those audiosegments and trying to line them up, you took a totally different approach. (Note: It was *theoretically* possible to do this using 'silence' and having two tracks playing at the same time, and while one track is playing music, the other is playing silence, and when the song is ending, the one that was playing silence begins to play it's song. I made an illus
tration on how it works)
![crossfade look](https://user-images.githubusercontent.com/102321690/188244038-c52b7b5a-dc4b-454a-ad27-795f93102fcd.png)

With no time.sleep(), it, (again, in theory) seems a lot easier to implement crossfade. Can't say for sure without spending a few dozen hours perfecting it though

Different operations, such as pausing, scrubbing (or parsing) though the song, resuming, all seem a little bit more difficult with this implementation, making sure that with each pause / play (especially on longer tracks, where more pauses could = some type of desync) lines up successfully, without being off at all.

This would also totally be more resource intensive than no crossfade. I'm unsure if pydub's silence occupies the same amount of memory as regular AudioSegments, but if so, a 15 minute video could easily break 300mb of memory, which IMO, is waaay too much

**Reply by tikuma-lsuhsc** [reply]
I think what you're facing is the limitation of `simpleaudio`. It is called `simple` for a reason. :-D

Have you considered `pyaudio`? `pyaudio`'s stream class will give you much more flexibility and will massively reduce the memory footprint.

**Comment by tikuma-lsuhsc** [comment]
Here is a question on `Punge`. Does it play a fixed playlist from the beginning to the end?

**Reply by nednoodlehead** [reply]
No, you have a list, the list contains objects. The objects represent a song. Each object has the following attributes:
>Title
>Author
>Album
>SavelocationThumb 
>Savelocation
>Unique id

The list is derived from sqlite3 tables, which are basically playlists. So a user can create playlists, which is essentially a 1:1 recreation of the table, if that makes sense.

Unique id is the primary key for the database
Savelocations are the locations on the user's device where the song / thumbnail is saved

The music class is used to move around this list. For shuffle, it random.shuffle(my_list), then has a reassemble function to put it back together. You can also specifically play the song of your choice, skip forwards & backwards. Basically all of the features you'd expect from a music app besides the ability to scrub through a song.

If you look briefly at the class attributes for 'music_player' (line 447), you can see all of the attributes, and what they do.
It is used by creating one instance of the music player class on app startup, then changing it while using the app

**Reply by tikuma-lsuhsc** [reply]
OK, I'm getting a better sense of how crossfade is being implemented in `pydub` (and I see how inefficient `pydub` is). 

Am I correct that it is not doing true seemless crossfading? There must be a split second between 2 songs.

I think you can do true crossfading with `ffmpegio`+`pyaudio` although you must implement the crossfading mechanism yourself:

1. add fade-in and fade-out filters to `ffmpegio` reader
2. add previous fade-out samples + next fade-in samples before queuing the overlapped samples to the `pyaudio` player

If this makes sense to you...

**Reply by nednoodlehead** [reply]
To be fair to pydub, it wasn't exactly meant to be used for playing audio in this sense. It has a lot of functionality for analyzing and processing audio, rather than more 'personal' playing music.

It does have true seamless crossfade, but only for a select amount of audio segments. Basically: Song1 + Song2 = (Song1 [crossfade here] Song2). <- Being one audiosegment.
Again, the problem is chaining a bunch of them together for full seamlessness. You'd need to turn that 2 audiosegment -> 10, or 100 song audiosegment for *true* seamlessness.

I think having more control over the crossfade, like you mention, directly editing the samples and whatnot will probably provide a much nicer experience than using and interface, which uses and interface that uses ffmpeg.

Again, a bit over my head, but im sure when using it, it'll come easy

**Comment by tikuma-lsuhsc** [comment]
Here is `pyaudio` example:

```python

# playback configuration (shared by both ffmpegio and pyaudio)
ar = 44100  # playback sampling rate
ac = 2  # number of channels
width = 2  # signed 2-byte integer format
sample_fmt = "s16"

# open ffmpegio's stream-reader
with ffmpegio.open(file, "ra", sample_fmt=sample_fmt, ac=ac, ar=ar) as f:

    # define pyaudio callback
    def callback(_, nblk, *__):
        # read nblk samples from the ffmpeg and pass it onto pyaudio
        data = f.read(nblk)["buffer"]
        return (data, pyaudio.paContinue)

    # start pyaudio stream
    p = pyaudio.PyAudio()
    format = p.get_format_from_width(width, unsigned)

    stream = p.open(rate, channels, format, output=True, stream_callback=callback,)
    stream.start_stream()

    # wait for stream to finish
    while stream.is_active():
        sleep(0.1)

    stream.stop_stream()
    stream.close()
    p.terminate()
```

You can find a working example in the [`python-ffmpegio-example-audioplayer` repo](https://github.com/python-ffmpegio/python-ffmpegio-example-audioplayer/blob/master/pyaudio_player.py).

**Comment by tikuma-lsuhsc** [comment]
Also got the cross-fading to work with `pyaudio`. The code is too long to post here, but go take a look [here](https://github.com/python-ffmpegio/python-ffmpegio-example-audioplayer/blob/master/pyaudio_player_crossfade.py).

Here are some keys:

`ffmpeg` is configured with a pair of `afade` filters (plus `aformat` to preformat the stream to the output format)

`ffmpegio.open` is called from `file_reader()` thread:

* processes a block of data, which matches the length of the fading (`tfade` in seconds, `nfade` in samples, and `nblk` in bytes)
* first block of the file is mixed with the last block of the previous file (both with fading effects created by FFmpeg)
* the block of samples are queued to a double-buffer, which gets read by the `pyaudio` callback

`pyaudio`'s thread calls `pyaudio_callback()`:

* `pyaudio` dictates the size of data block to return, which do not match the data block in the queue (`nblk` bytes)
* `buf` stores whatever is left over from the previous queue access, gets the next block when `buf` is exhausted

If you have any questions, ask here (paste the code lines in question)

**Reply by nednoodlehead** [reply]
On 'testfile_generator.py', what exactly is it yielding when returning?
yield [generate(i) for i in range(nfiles)]

And I guess this ties into that, but right after the context manager in 'pyaudio_player.py':
with testfiles(1) as files:
    file = files[0]

Why does we only use the [0] of the testfiles(1)? What is it's purpose?

Also the sleep(0.1):

            # wait for stream to finish
            while stream.is_active():
                sleep(0.1)

What is the purpose of that? I removed the time.sleep and it acted fine. I'm assuming with this information, this has nothing to do with the time.sleep() stuff i was talking about with how my program works?

**Reply by tikuma-lsuhsc** [reply]
> On 'testfile_generator.py', what exactly is it yielding when returning?
> `yield [generate(i) for i in range(nfiles)]`

Paths to temporally audio files it generates. Once your main execution goes out of `with` scope, all these temp files get deleted. Do `print (files)` in the main script to see what it looks like.

>And I guess this ties into that, but right after the context manager in 'pyaudio_player.py':
>
>```
>with testfiles(1) as files:
>    file = files[0]
>```
>
>Why does we only use the [0] of the testfiles(1)? What is it's purpose?

As the generator function name `testfiles` suggests, it always returns a list of filepaths.  On the plain pyaudio player script, it can only play 1 audio file, so it's generating 1 file (the first argument=`nfiles`) and only using the first and only entry.

>Also the `sleep(0.1)`:
>
>```
># wait for stream to finish
>while stream.is_active():
>    sleep(0.1)
>```
>
>What is the purpose of that? I removed the time.sleep and it acted fine. I'm assuming with this information, this has nothing to do with the time.sleep() stuff i was talking about with how my program works?

It's the same. Without `sleep` your main thread stays busy constantly checking the state of the player, needlessly burning PC resources. Ideally, you'd want to use `thread.join()` or `process.wait()` to wait for another thread/process, but `pyaudio` does not expose any method like that. I may deep-dive `portaudio` source in the future to see if a feature like that can be added.

Meanwhile, keep your `sleep()` lines intact.

**Reply by nednoodlehead** [reply]
Yeah I see what you mean with the testfiles. I figured out how to play my own song fairly easily. Just setting the ffmpegio.open()'s first value to the path to my song.

Also, is there any native way to get the time passed in the song? That was one of the problems I had. I actually had a very similar implementation, that didn't get added to my app because I sort of feared that constantly updating the time left / time passed in a song would generate significant overhead. Yes, you could use flags or whatever, but I hadn't gotten that far.

Also, have you gotten to test the memory usage for playing the files? I would love the flexibility of not struggling to play long (40m+) videos without wondering if it has a memory leak or something. 

I also had the problem, that I sort of had to band-aid fix; where long videos would take a fair bit of time to initialize, and if someone clicked play (begins to init long video), they could click again, and cause some playback errors. I had to fix it by putting a cooldown on the button. I'm sure if we get this project up and running, it should likely solve that issue.

**Reply by tikuma-lsuhsc** [reply]
> Also, is there any native way to get the time passed in the song? 

This is the major shortcoming of both `simpleaudio` and `pyaudio`. They offer so little to monitor the state of audio playback. As I mentioned below, Qt's [QAudioSink](https://doc.qt.io/qt-6/qaudiosink.html) opens up a possibility to track the seconds played [`elapsedUSecs()`](https://doc.qt.io/qt-6/qaudiosink.html#elapsedUSecs). You may want to look into the world of `PyQt6/5`

> Also, have you gotten to test the memory usage for playing the files? 

No. But it shouldn't drive up memory usage at all as the buffer size is fixed throughout.

Have you had chance to study the crossfade example?

**Reply by nednoodlehead** [reply]
I have not had a chance to thoroughly look at the crossfade example. I was mostly poking around with the more basic example is 'pyaudio_player.py', as to understand a bit more of this 'low level' audio manipulation. I'm really trying to build the solid foundation of understanding the regular player, in and out, before moving on to the crossfade.

In the context of my project, crossfade will not be implemented for a little while. It is just so dang hard to implement. Perhaps, at it's minimum, crossfade is not difficult, as you showed me in the example. However, incorporating traditional media player functions into that mix is a different story. For example, skipping a song. Skipping a song with the regular class I'm using right now in Punge works well. You increase the song_count by 1 and move on. Throwing crossfade into that might be a new can of worms in terms of bugs. Where do you skip to exactly? Before the crossfade has fully finished or the end of it? Do you start a whole new 'instance' all together where there is no crossfade in the beginning?

Would I like to have crossfade? Of course. Is it (going to be) difficult? Probably.

**Reply by tikuma-lsuhsc** [reply]
> I have not had a chance to thoroughly look at the crossfade example. I was mostly poking around with the more basic example is 'pyaudio_player.py', as to understand a bit more of this 'low level' audio manipulation. I'm really trying to build the solid foundation of understanding the regular player, in and out, before moving on to the crossfade.

That's a sound idea. The complexity of the crossfade example is much much higher in terms of how you pass around a buffer of sound samples from thread to thread. Once you understand how `pyaudio`'s callback works, you're 90% of the way there IMO.

> Throwing crossfade into that might be a new can of worms in terms of bugs. Where do you skip to exactly? Before the crossfade has fully finished or the end of it? Do you start a whole new 'instance' all together where there is no crossfade in the beginning?

Crossfade and skipping are 2 completely different animals at least in how my demo is written... actually to skip a song, you do need to stop and restart (`pyaudio` doesn't let you clear the buffer during playback). Meanwhile, crossfade is automatically inserted between tracks.

> Would I like to have crossfade? Of course. Is it (going to be) difficult? Probably.

With `pydub` without a question, it would be hard. Once you get used to buffered "realtime" playback, you may find it much easier.

**Comment by nednoodlehead** [comment]
Do you want this media player bit I'm gunna create to be apart of this repo or should we have a separate one? I'm fine with it in here, but it you want to separate it, no biggie.

**Reply by tikuma-lsuhsc** [reply]
Actually, I currently don't have any interest in creating a player. I'm using this opportunity to learn how to interface FFmpeg with various audio player packages. My hope was that you can go back to your `Punge` and incorporate `ffmpegio` to it. I'd be happy to help.

Oh, another thing... If you ever want to learn new UI framework, look into Qt. It has control interaface for [audio device](https://doc.qt.io/qt-6/qaudiosink.html). I'll give it a try next.

**Reply by nednoodlehead** [reply]
I actually did have plans to use Qt5 to re-do my GUI, I've just been super lazy in doing so. I'm assuming the docs you linked also has a python equivalent because those are for c++ lol.

**Reply by tikuma-lsuhsc** [reply]
My limited experience with PyQt, I found C++ documentation is all you need as everything is essentially there as is in Python (ofc only if you are proficient in C++). (Tricky ones being find the constants.) No idea if there is PyQt counter part to `QMultimedia` documentation.

**Comment by nednoodlehead** [comment]
Hey, just seeing where this project is at, have you made any type of media-type player like we were talking about? My current goal is to learn a bit of PyQt5 (in progress :) ), then revamp my [Punge](https://github.com/nednoodlehead/punge) project with PyQt5 and this project.

Of course it'll take me a few weeks to get a working ffmpegio up and going how i want it, but, hopefully soon !
